mail.go

```diff
@@ -0,0 +1,190 @@
+)
+
+const (
+	FileMimeType           = "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
+	FileContentDisposition = "attachment"
+)
+
+func setHeader(headers *textproto.MIMEHeader, key, value string) {
+	h := textproto.CanonicalMIMEHeaderKey(key)
+	headers.Set(h, value)
+}
+
+func addHeader(headers *textproto.MIMEHeader, key, value string) {
+	h := textproto.CanonicalMIMEHeaderKey(key)
+	headers.Add(h, value)
+}
+
+// create a multipart MIME message with text and file attachment
+func setBody(headers *textproto.MIMEHeader, body *bytes.Buffer, f *excelize.File) error {
+
+	multipartType := "fixed"
+
+	// create a new multipart writer
+	writer := multipart.NewWriter(body)
+	contentType := "multipart/" + multipartType + ";\n \tboundary=" + writer.Boundary()
+	// add header to main header group
+	setHeader(headers, "Content-Type", contentType)
+
+	// add text part
+	textBody := "\r\nPlease see attached file with portal accounts.\r\n"
+	textHeader := make(textproto.MIMEHeader)
+	setHeader(&textHeader, "Content-Type", "text/plain; charset="+charSet)
+	setHeader(&textHeader, "Content-Transfer-Encoding", bodyEncoding)
+	p, err := writer.CreatePart(textHeader)
+	if err != nil {
+		return err
+	}
+	_, err = p.Write([]byte(b64.StdEncoding.EncodeToString([]byte(textBody))))
+	if err != nil {
+		return err
+	}
+
+	// add file attachment part
+	fp, err := os.Open(f.Path)
+	if err != nil {
+		return err
+	}
+	defer fp.Close()
+
+	data, err := io.ReadAll(fp)
+	if err != nil {
+		return err
+	}
+
+	fileHeader := make(textproto.MIMEHeader)
+	value := fmt.Sprintf("%s;\n \tname=%q", FileMimeType, AttachedFileName)
+	setHeader(&fileHeader, "Content-Type", value)
+	setHeader(&fileHeader, "Content-Transfer-Encoding", bodyEncoding)
+	value = fmt.Sprintf("attachment;\n \tfilename=%q", AttachedFileName)
+	setHeader(&fileHeader, "Content-Disposition", value)
+	p, err = writer.CreatePart(fileHeader)
+	if err != nil {
+		return err
+	}
+	_, err = p.Write([]byte(b64.StdEncoding.EncodeToString(data)))
+	if err != nil {
+		return err
+	}
+
+	err = writer.Close()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func toMIMEString(headers *textproto.MIMEHeader, body *bytes.Buffer) string {
+	return toHeaderString(headers) + body.String()
+}
+
+func toHeaderString(headers *textproto.MIMEHeader) string {
+	setHeader(headers, "Date", time.Now().Format(time.RFC1123Z))
+
+	// combine the headers
+	strHeader := ""
+	for header, values := range *headers {
+		strHeader += header + ": " + strings.Join(values, ", ") + "\r\n"
+	}
+
+	return strHeader + "\r\n"
+}
+
+func validateRecipientAddresses(addresses []string) ([]string, error) {
+	// remove duplicate addresses
+	// log invalid address and continue
+	// if no valid address, set m.err
+	validAddresses := make([]string, 0)
+	for _, addr := range addresses {
+		if len(addr) > 0 {
+			a, err := mail.ParseAddress(addr)
+			if err != nil {
+				log.Printf("validating recipient address %s: %s", addr, err)
```

Comment:
This is not something that could happen only occasionally or randomly, right? What do you think about failing so that the user knows they need to fix the issue?

Reply:
I have 2 log filters, each of which has an associated alarm: `password-rotation-info` and `password-rotation-errors`.  Validation errors that don't cause the program to stop count towards the `password-rotation-info` alarm. Both alarms send a notification to the same SNS topic, which sends an email to the testing team. However, I need to fix the `info` filter to catch this log entry. (I checked the `info` filter and it does not filter on "validating", as I incorrectly thought it did. 

I will add `Info: ` at the beginning of this log entry and make a new PR that fixes the `info` filter to filter on the word `Info` and adds `Info` to the other log entries that current match the `info` filter. Is that OK?

Reply:
Sounds good, just wanted to make sure it wasn't an oversight.